src/runtime_src/core/tools/common/ReportAsyncError.cpp

```diff
@@ -0,0 +1,128 @@
+/**
+ * Copyright (C) 2020 Xilinx, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+// ------ I N C L U D E   F I L E S -------------------------------------------
+// Local - Include Files
+#include <ctime>
+#include "ReportAsyncError.h"
+#include "core/common/query_requests.h"
+#include "core/common/device.h"
+#include "core/include/xrt.h"
+#include "core/include/experimental/xrt_error.h"
+#include "core/include/xrt_error_code.h"
+#include <boost/algorithm/string.hpp>
+#include <boost/property_tree/json_parser.hpp>
+namespace qr = xrt_core::query;
+
+/**
+ * get time in nano second
+ */
+std::string time(long unsigned data)
+{
+  time_t rdata = data/1000000000;
+  struct tm *timeptr = std::localtime(&rdata);
+  static const char mon_name[][4] = {
+    "Jan", "Feb", "Mar", "Apr", "May", "Jun",
+    "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
+  };
+  int year = (1900 + timeptr->tm_year); 
+  int nsec = (data%1000000000); 
+  return boost::str(boost::format("%.3s%3d %.2d:%.2d:%.2d:%.9d %d")
+    % mon_name[timeptr->tm_mon]
+    % timeptr->tm_mday % timeptr->tm_hour
+    % timeptr->tm_min % timeptr->tm_sec
+    % nsec % year);
+}
+
+boost::property_tree::ptree
+populate_async_error(const xrt_core::device * device)
+{
+  boost::property_tree::ptree pt;
+  boost::property_tree::ptree error_array;
+  auto dhdl = xrtDeviceOpenFromXcl(device->get_device_handle());
+  for (xrtErrorClass ecl = XRT_ERROR_CLASS_FIRST_ENTRY; ecl < XRT_ERROR_CLASS_UNKNOWN; ecl = xrtErrorClass(ecl+1)) {
+    xrtErrorCode errorCode;
+    uint64_t timestamp;
+    int rval = xrtErrorGetLast(dhdl, ecl, &errorCode, &timestamp);
+    /**
+     * In case of no error for given class, errorCode and timestamp will be zero,
+     * but zero errorCode is valid error, so checking for only timestamp.
+     */
+    if (rval == 0 && timestamp) {
+      size_t len = 0;
+      if (xrtErrorGetJson(dhdl, errorCode, nullptr, 0, &len))
+        continue;
+      char buf[len];
+      if (xrtErrorGetJson(dhdl, errorCode, buf, len, nullptr))
+        continue;
+      std::stringstream ss(buf);
+      boost::property_tree::ptree _pt;
+      boost::property_tree::read_json(ss, _pt);

```

Comment:
It seems a little strange that we create the property tree in the xrt_error.cpp, convert it to string and return it, only to recreate the property tree here.   I don't think the API in xrt_error.h should return a property tree, but I do think that you could all an internal API in core/common/api/error_int.h (see other such headers) where you provide access to error_code_to_json() and have this function return the property tree rather than string.  Then in the public API you convert the property tree to string for end-user.

Reply:
Agree...comments added above indicate that this "conversion" layer should be removed.

Reply:
changed

Reply:
Thank you.